protocol/x/ratelimit/keeper/keeper.go

```diff
@@ -218,6 +221,159 @@ func (k Keeper) SetDenomCapacity(
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"denom (%s) capacity list length (%v) != limiters length (%v); skipping capacity update",
+				limitParams.Denom,
+				len(capacityList),
+				len(limitParams.Limiters),
+			),
+		)
+		return
+	}
+
+	// Convert timestamps to milliseconds for algebraic operations.
+	blockTimeMilli := ctx.BlockTime().UnixMilli()
+	prevBlockInfo := k.blockTimeKeeper.GetPreviousBlockInfo(ctx)
+	prevBlockTimeMilli := prevBlockInfo.Timestamp.UnixMilli()
+	timeSinceLastBlockMilli := new(big.Int).Sub(
+		big.NewInt(blockTimeMilli),
+		big.NewInt(prevBlockTimeMilli),
+	)
+	if timeSinceLastBlockMilli.Sign() <= 0 {
+		// This violates an invariant (current block time > prev block time).
+		// Since this is in the `EndBlocker`, we log an error instead of panicking.
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"timeSinceLastBlockMilli (%v) <= 0; skipping capacity update. prevBlockTimeMilli = %v, blockTimeMilli = %v",
+				timeSinceLastBlockMilli,
+				prevBlockTimeMilli,
+				blockTimeMilli,
+			),
+		)
+		return
+	}
+
+	// Declare new capacity list to be populated.
+	newCapacityList := make([]dtypes.SerializableInt, len(capacityList))
+
+	for i, limiter := range limitParams.Limiters {
+		// For each limiter, calculate the current baseline.
+		baseline := k.GetBaseline(ctx, supply.Amount.BigInt(), limiter)
+
+		capacityMinusBaseline := new(big.Int).Sub(
+			capacityList[i].BigInt(), // array access is safe because of the invariant check above
+			baseline,
+		)
+
+		// Calculate left operand: `max(baseline, capacity-baseline)`. This equals `baseline` when `capacity <= 2 * baseline`
+		operandL := new(big.Rat).SetInt(
+			lib.BigMax(
+				baseline,
+				capacityMinusBaseline,
+			),
+		)
+
+		// Calculate right operand: `time_since_last_block / period`
+		periodMilli := new(big.Int).Mul(
+			new(big.Int).SetUint64(uint64(limiter.PeriodSec)),
+			big.NewInt(1000),
+		)
+		operandR := new(big.Rat).Quo(
+			new(big.Rat).SetInt(timeSinceLastBlockMilli),
+			new(big.Rat).SetInt(periodMilli),
+		)
+
+		// Calculate: `capacity_diff = max(baseline, capacity-baseline) * (time_since_last_block / period)`
+		// Since both operands > 0, `capacity_diff` is positive or zero (due to rounding).
+		capacityDiff := new(big.Rat).Mul(
+			operandL,
+			operandR,
+		)
+
+		bigRatcapacityMinusBaseline := new(big.Rat).SetInt(capacityMinusBaseline)
+
+		if new(big.Rat).Abs(bigRatcapacityMinusBaseline).Cmp(capacityDiff) < 0 {
+			// if `abs(capacity - baseline) < capacity_diff` then `capacity = baseline``
+			newCapacityList[i] = dtypes.NewIntFromBigInt(baseline)
+		} else if capacityList[i].BigInt().Cmp(baseline) < 0 {
+			// else if `capacity < baseline` then `capacity += capacity_diff`
+			newCapacityList[i] = dtypes.NewIntFromBigInt(
+				new(big.Int).Add(
+					capacityList[i].BigInt(),
+					lib.BigRatRound(capacityDiff, false), // rounds down `capacity_diff`
```

Comment:
nit: didnt look at the unit tests closely, can you confirm that we have unit tests that verifies the round down behaviors for this method?

Reply:
Yes wo do!